File: linked-list-cycle-ii/linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        //check for null input 
        if(head == null) return head;
        ListNode slow = head;
        ListNode fast = head;
        ListNode tmp = head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast) break;
        }
        if(fast == null || fast.next == null)return null;
        while(tmp != slow){
            slow = slow.next;
            tmp = tmp.next;
        }
        return tmp;
    }
}
//Time complexity:O(n)
//space complexity:O(1)

File: linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None: return head;
        slow = head
        fast = head
        tmp = head
        while(fast is not None and fast.next is not None):
            slow = slow.next
            fast = fast.next.next
            if slow == fast: break
        if fast is None or fast.next is None: return None
        while tmp != slow:
            tmp = tmp.next
            slow = slow.next
        return tmp